### ISSUE TITLE
.net xmlrpc-api breaks when avp has empty key

### ISSUE THREAD

**Issue by SoundGoof** [SpriteLink/NIPAP#876]
the solution perhaps is to restrict creation of empty key in avp?
it is very easy in the gui to click add avp and then go to save and then save (this is what is happening for me now(my users))

errormessage from .net: CookComputing.XmlRpc.XmlRpcInvalidXmlRpcException: Struct contains member with empty name element.

also in nipap-cli ony the latest avp with the same key-name is shown, but empty key works

**Comment by SoundGoof** [comment]
also in nipap-cli ony the latest avp with the same key-name is shown

**Comment by plajjan** [comment]
Yeah, so having an empty AVP (ie name is '') doesn't make any sense whatsoever so I suppose we should filter that out in the backend. I don't think we need to return an error or anything, we can just silently ignore it. That would mean the web UI doesn't need any modifications (although I think it would be cleaner not to send AVP with name ''.

It also means backend never stores empty AVP and subsequently it never returns an empty AVP and therefore any issue with the CLI is automatically solved.

@garberg waddya think of silently ignoring an AVP with an empty name?

**Comment by garberg** [comment]
I'm not a big fan of the idea of throwing data sent to the backend away without notifying the client, but it could be OK to discard in the web UI?

**Comment by plajjan** [comment]
@garberg Right, I'm with you over all, which is why I'm asking you, but it's an AVP with empty name that we are talking about here.. do people really want to add that? Sounds more like a mistake.

I want to fix in the web UI as well but that would not constitute a fix to this issue. The problem is .net xml-rpc framework receiving empty name element. I don't want us to rely on just good behaving clients to avoid such issues, that feels quite naive, which is why I want to go touch the backend.

We could of course let web UI & CLI drop empty attributes and in the backend we return an error, so it's not silent.

**Comment by garberg** [comment]
Yeah, sorry, the last paragraph was exactly my idea which did not really come out in my repsonse :P

In the web UI and CLI, drop attributes without a name (Or perhaps notify the user if a value has been filled in?). In the backend, return an error.

**Comment by plajjan** [comment]
I don't think we should notify the user, at least not if both name and value is empty because that probably just means that the user pushed "add attribute" button and then changed their mind. An error message is just annoying for that, no?

If there is a value for an AVP with an empty name though, that's slightly different, maybe someone accidentally removed the name? Not sure how to handle.

**Comment by plajjan** [comment]
We are in agreement on what the backend should do though so we can go ahead and implement that. Let's start with a unit test.

**Comment by plajjan** [comment]
So I opened #881 for the backend and #882 for the web UI. I don't actually think this is a problem for the CLI as it will print the error from the backend so it should be fixed with #881. We do need to figure out what to do with existing data though.

**Comment by plajjan** [comment]
CLI returns with error:
```
kll@lingloiVM ~/kod/NIPAP/tests $ nipap address add prefix 1.2.3.0/24 description foo type assignment extra-attribute =1337
Could not add prefix to NIPAP: AVP with empty name is not allowed
kll@lingloiVM ~/kod/NIPAP/tests
```
I think this is good enough.

**Comment by plajjan** [comment]
@SoundGoof so I guess your issue is solved through the merge of #883, right? Can you please confirm and close this issue?

**Comment by SoundGoof** [comment]
this solves my issue ;)
